pkg/collector/corechecks/snmp/internal/report/report_memory_usage.go

```diff
@@ -0,0 +1,300 @@
+					options:    checkconfig.MetricsConfigOption{},
+					forcedType: "",
+				}
+
+				ms.sendMetric(memoryUsageSample)
+			}
+			return nil
+		}
+
+		if memoryFreeOk {
+			for rowIndex, memoryUsedSample := range memoryUsedRows {
+				memoryFreeSample, memoryFreeSampleOk := memoryFreeRows[rowIndex]
+				if !memoryFreeSampleOk {
+					return fmt.Errorf("missing memory free sample at row %s, skipping memory usage evaluation", rowIndex)
+				}
+				floatMemoryFree, err := memoryFreeSample.value.ToFloat64()
+				if err != nil {
+					return fmt.Errorf("metric `%s[%s]`: failed to convert to float64: %s", "memory.free", rowIndex, err)
+				}
+
+				floatMemoryUsed, err := memoryUsedSample.value.ToFloat64()
+				if err != nil {
+					return fmt.Errorf("metric `%s[%s]`: failed to convert to float64: %s", "memory.used", rowIndex, err)
+				}
+
+				memoryUsageValue, err := evaluateMemoryUsage(floatMemoryUsed, floatMemoryFree+floatMemoryUsed)
+				if err != nil {
+					return err
+				}
+
+				memoryUsageSample := MetricSample{
+					value:      valuestore.ResultValue{Value: memoryUsageValue},
+					tags:       memoryUsedSample.tags,
+					symbol:     checkconfig.SymbolConfig{Name: "memory.usage"},
+					options:    checkconfig.MetricsConfigOption{},
+					forcedType: "",
+				}
+
+				ms.sendMetric(memoryUsageSample)
+			}
+			return nil
+		}
+	}
+
+	if memoryFreeOk && memoryTotalOk {
+		for rowIndex, memoryTotalSample := range memoryTotalRows {
+			memoryFreeSample, memoryFreeSampleOk := memoryFreeRows[rowIndex]
+			if !memoryFreeSampleOk {
+				return fmt.Errorf("missing memory free sample at row %s, skipping memory usage evaluation", rowIndex)
+			}
+			floatMemoryFree, err := memoryFreeSample.value.ToFloat64()
+			if err != nil {
+				return fmt.Errorf("metric `%s[%s]`: failed to convert to float64: %s", "memory.free", rowIndex, err)
+			}
+
+			floatMemoryTotal, err := memoryTotalSample.value.ToFloat64()
+			if err != nil {
+				return fmt.Errorf("metric `%s[%s]`: failed to convert to float64: %s", "memory.total", rowIndex, err)
+			}
+
+			memoryUsageValue, err := evaluateMemoryUsage(floatMemoryTotal-floatMemoryFree, floatMemoryTotal)
+			if err != nil {
+				return err
+			}
+
+			memoryUsageSample := MetricSample{
+				value:      valuestore.ResultValue{Value: memoryUsageValue},
+				tags:       memoryTotalSample.tags,
+				symbol:     checkconfig.SymbolConfig{Name: "memory.usage"},
+				options:    checkconfig.MetricsConfigOption{},
+				forcedType: "",
+			}
+
+			ms.sendMetric(memoryUsageSample)
+		}
+		return nil
+	}
+
+	// report missing dependency metrics
+	missingMetrics := []string{}
+	if !memoryUsedOk {
+		missingMetrics = append(missingMetrics, "used")
+	}
+	if !memoryFreeOk {
+		missingMetrics = append(missingMetrics, "free")
+	}
+	if !memoryTotalOk {
+		missingMetrics = append(missingMetrics, "total")
+	}
+
+	return fmt.Errorf("missing %s memory metrics, skipping column memory usage", strings.Join(missingMetrics, ", "))
```

Comment:
Shall factor this with similar code in column impl ?

Reply:
I could refactor using a map of metric names, and check for each if it is missing. But I honestly like it this way, makes it more readable, again opinionated.

Reply:
Sorry, if I wasn't precise enough. I'm referring to the duplication of this section at those 2 places:
- https://github.com/DataDog/datadog-agent/pull/11993/files/0b0fd6198edfa2e60b87fa4f2f0a2c0238eb9e5f#diff-50ccdd9110d345434c7d85be3b7abb474fa2dd60927578eecd04987944aeaa5dR277-R289
- https://github.com/DataDog/datadog-agent/pull/11993/files/0b0fd6198edfa2e60b87fa4f2f0a2c0238eb9e5f#diff-50ccdd9110d345434c7d85be3b7abb474fa2dd60927578eecd04987944aeaa5dR147-R159